cpu/gd32v/periph/gpio_ll_irq.c

```diff
@@ -0,0 +1,192 @@
+ * @ingroup     drivers_periph_gpio_ll_irq
+ * @{
+ *
+ * @file
+ * @brief       IRQ implementation of the GPIO Low-Level API for GD32V
+ *
+ * @author      Gunar Schorcht <gunar@schorcht.net>
+ * @}
+ */
+
+#include <errno.h>
+
+#include "cpu.h"
+#include "bitarithm.h"
+#include "periph/gpio_ll_irq.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+/**
+ * @brief   Number of available external interrupt lines
+ */
+#define GPIO_ISR_CHAN_NUMOF             (16U)
+#define GPIO_ISR_CHAN_MASK              (0xFFFF)
+
+struct _gpio_isr_ctx {
+    gpio_ll_cb_t cb;
+    void *arg;
+};
+
+/**
+ * @brief   Allocate memory for one callback and argument per EXTI channel
+ */
+static struct _gpio_isr_ctx _exti_ctx[GPIO_ISR_CHAN_NUMOF];
+
+/* keep the state of level triggered interrupt activation using 32-bit values
+ * since the registers require 32-bit access */
+static uint32_t _h_level_triggered = 0;
+static uint32_t _l_level_triggered = 0;
+
+void gpio_ll_irq_mask(gpio_port_t port, uint8_t pin)
+{
+    (void)port;
+    EXTI->INTEN &= ~(1 << pin);
+}
+
+void gpio_ll_irq_unmask_and_clear(gpio_port_t port, uint8_t pin)
+{
+    (void)port;
+    EXTI->PD = (1 << pin);
+    EXTI->INTEN |= (1 << pin);
+}
+
+static inline unsigned _irq_num(unsigned pin)
+{
+    if (pin < 5) {
+        return EXTI0_IRQn + pin;
+    }
+    if (pin < 10) {
+        return EXTI5_9_IRQn;
+    }
+    return EXTI10_15_IRQn;
+}
+
+/* Forward declaration of ISR */
+static void _gpio_isr(unsigned irqn);
+
+int gpio_ll_irq(gpio_port_t port, uint8_t pin, gpio_irq_trig_t trig, gpio_ll_cb_t cb, void *arg)
+{
+    unsigned irq_state = irq_disable();
+    int port_num = GPIO_PORT_NUM(port);
+
+    /* set callback */
+    _exti_ctx[pin].cb = cb;
+    _exti_ctx[pin].arg = arg;
+
+    /* enable alternate function clock for the GPIO module */
+    periph_clk_en(APB2, RCU_APB2EN_AFEN_Msk);
+
+    /* configure the EXTI channel */
+    volatile uint32_t *afio_exti_ss = &AFIO->EXTISS0 + (pin >> 2);
+
+    *afio_exti_ss &= ~(0xfUL << ((pin & 0x03) * 4));
+    *afio_exti_ss |= (uint32_t)port_num << ((pin & 0x03) * 4);
+
+    uint32_t pin_mask = 1UL << pin;
+
+    _h_level_triggered &= ~pin_mask;
+    _l_level_triggered &= ~pin_mask;
+    EXTI->RTEN &= ~pin_mask;
+    EXTI->FTEN &= ~pin_mask;
+
+    /* configure the active flank, level interrupts are emulated */
+    switch (trig) {
+        case GPIO_TRIGGER_EDGE_RISING:
+            EXTI->RTEN |= pin_mask;
+            break;
+        case GPIO_TRIGGER_EDGE_FALLING:
+            EXTI->FTEN |= pin_mask;
+            break;
+        case GPIO_TRIGGER_EDGE_BOTH:
+            EXTI->RTEN |= pin_mask;
+            EXTI->FTEN |= pin_mask;
+            break;
+        case GPIO_TRIGGER_LEVEL_HIGH:
+            EXTI->RTEN |= pin_mask;
+            _h_level_triggered |= pin_mask;
+            break;
+        case GPIO_TRIGGER_LEVEL_LOW:
+            EXTI->FTEN |= pin_mask;
+            _l_level_triggered |= pin_mask;
+            break;
+    }
```

Comment:
```suggestion
    switch (trig) {
    case GPIO_TRIGGER_EDGE_RISING:
        EXTI->RTEN |= pin_mask;
        break;
    case GPIO_TRIGGER_EDGE_FALLING:
        EXTI->FTEN |= pin_mask;
        break;
    case GPIO_TRIGGER_EDGE_BOTH:
        EXTI->RTEN |= pin_mask;
        EXTI->FTEN |= pin_mask;
        break;
    case GPIO_TRIGGER_LEVEL_HIGH:
        EXTI->RTEN |= pin_mask;
        _h_level_triggered |= pin_mask;
        break;
    case GPIO_TRIGGER_LEVEL_LOW:
        EXTI->FTEN |= pin_mask;
        _l_level_triggered |= pin_mask;
        break;
    }
```

Reply:
When I wrote this switch case construct, I knew you were going to ask for a change in style 😉 I still don't find it that readable. I will change andsquah it directly.